src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,72 +9,316 @@
     rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
+    MPIDI_XPMEM_REQUEST(rreq, counter_ptr) = NULL;
 
-    /* Send ack to sender */
-    slmt_ack_hdr->sreq_ptr = sreq_ptr;
     mpi_errno =
-        MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
-                               MPIDIG_context_id_to_comm(MPIDIG_REQUEST(rreq, context_id)),
-                               MPIDI_SHM_XPMEM_SEND_LMT_ACK, &ack_ctrl_hdr);
+        MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset, &seg_ptr, &src_buf,
+                               &MPIDI_XPMEM_global.segmaps[src_lrank].segcache_user);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    XPMEM_TRACE("handle_lmt_single_recv: handle matched rreq %p [source %d, tag %d, "
+                " context_id 0x%x], copy dst %p, src %lx, bytes %ld\n", rreq,
+                MPIDIG_REQUEST(rreq, rank), MPIDIG_REQUEST(rreq, tag),
+                MPIDIG_REQUEST(rreq, context_id), (char *) MPIDIG_REQUEST(rreq, buffer),
+                src_offset, recv_data_sz);
+
+    /* Copy data to receive buffer */
+    mpi_errno = MPIR_Localcopy(src_buf, recv_data_sz, MPI_BYTE,
+                               MPIDIG_REQUEST(rreq, buffer), MPIDIG_REQUEST(rreq, count),
+                               MPIDIG_REQUEST(rreq, datatype));
+    MPIR_ERR_CHECK(mpi_errno);
+
+    ack_ctrl_hdr.xpmem_slmt_send_fin.req_ptr = sreq_ptr;
+    mpi_errno = MPIDI_SHM_do_ctrl_send(MPIDIG_REQUEST(rreq, rank),
+                                       MPIDIG_context_id_to_comm(MPIDIG_REQUEST
+                                                                 (rreq, context_id)),
+                                       MPIDI_SHM_XPMEM_SEND_LMT_RECV_FIN, &ack_ctrl_hdr);
     MPIR_ERR_CHECK(mpi_errno);
 
     MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
     MPID_Request_complete(rreq);
 
+    mpi_errno = MPIDI_XPMEM_seg_deregist(seg_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
   fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_SINGLE_RECV);
     return mpi_errno;
   fn_fail:
     goto fn_exit;
 }
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_recv(uint64_t src_offset, size_t data_sz,
+                                                         uint64_t sreq_ptr, int src_lrank,
+                                                         MPIR_Comm * comm, MPIR_Request * rreq)
+{
+
+    int mpi_errno = MPI_SUCCESS;
+    int recvtype_iscontig;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    /* TODO: need to support cooperative copy on non-contig datatype on receiver side. */
+    MPIR_Datatype_is_contig(MPIDIG_REQUEST(rreq, datatype), &recvtype_iscontig);
+    if (recvtype_iscontig && data_sz > MPIR_CVAR_CH4_XPMEM_COOP_COPY_THRESHOLD &&
+        src_lrank != MPIDI_XPMEM_global.local_rank) {
+        /* Cooperative XPMEM copy */
```

Comment:
Have you checked if self-message is handled in XPMEM?

Reply:
Yes, it handles self-message correctly.

Reply:
I actually do not think XPMEM should handle self-message. XPMEM is specifically used for large & inter-process message. `CH3` optimizes it in an upper layer but we might not include the code in CH4. Let us just fallback to POSIX for self-message in this PR (i.e., check if dest == myself, issue POSIX_send on sender side).